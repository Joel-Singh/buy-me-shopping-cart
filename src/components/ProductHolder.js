import StoreProduct from "./StoreProduct"
function ProductHolder({ productList, addToCartFunction, removeFromCartFunction }) {
  const productsAsHTML = productList.map(productInformation => {
    const name = productInformation.name

    return (
      <StoreProduct
        key={name}
        {...productInformation}
        onAddToCartClick={() => addToCartFunction(name)}
        onRemoveFromCartClick={() => removeFromCartFunction(name)}
      />)
  })

  return (
    <div>{productsAsHTML}</div>
  )
}

export default ProductHolder;
