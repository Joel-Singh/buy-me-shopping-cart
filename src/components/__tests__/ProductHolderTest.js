import ProductHolder from "../ProductHolder";
import { getByRole, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

test("Product holder properly renders a single product", () => {
  const productList = [
    {
      amountInCart: 5,
      image: "placeholder",
      name: "Me smiling",
    }
  ];

  const { container : productHolder } = render(<ProductHolder productList={productList} />)

  expect(productHolder).toMatchSnapshot()
})


test("Product holder properly renders three products", () => {
  const productList = [
    {
      amountInCart: 9,
      image: "frown placeholder",
      name: "Frown",
    },
    {
      amountInCart: 2,
      image: "smile placeholder",
      name: "Smile",
    },
    {
      amountInCart: 6,
      image: "depression placeholder",
      name: "Depression",
    },
  ];

  const { container : productHolder } = render(<ProductHolder productList={productList} />)

  expect(productHolder).toMatchSnapshot()
})

test("Cart functions of a single Product within a Product Holder are called when clicked", () => {
  const smileProduct = { name: "smile" }
  const productList = [
    smileProduct
  ];

  const addToCartFunction = jest.fn();
  const removeFromCartFunction = jest.fn();

  render(
    <ProductHolder
      productList={productList}
      addToCartFunction={addToCartFunction}
      removeFromCartFunction={removeFromCartFunction}
    />
  )

  const smileProductElement = screen.getByText(smileProduct.name).parentElement
  const addToCartButton = getByRole(smileProductElement, 'button', {name: "Add to cart"})
  const removeFromCartButton = getByRole(smileProductElement, 'button', {name: "Remove from cart"})

  userEvent.click(addToCartButton)
  userEvent.click(removeFromCartButton)

  expect(addToCartFunction).toBeCalledWith("smile")
  expect(removeFromCartFunction).toBeCalledWith("smile")
})

test("Cart functions of a multiple Products within a Product Holder are called when clicked", () => {
  const smileProduct = { name: "smile" }
  const frownProduct = { name: "frown" }
  const terrorProduct = {name: "terror" }

  const productList = [
    smileProduct,
    frownProduct,
    terrorProduct,
  ];

  const addToCartFunction = jest.fn();
  const removeFromCartFunction = jest.fn();

  render(
    <ProductHolder
      productList={productList}
      addToCartFunction={addToCartFunction}
      removeFromCartFunction={removeFromCartFunction}
    />
  )

  function getAddToCartButton(name) {
    const productElement = screen.getByText(name).parentElement
    const addToCartButton = getByRole(productElement, 'button', {name: "Add to cart"})
    return addToCartButton
  }

  function getRemoveFromCartButton(name) {
    const productElement = screen.getByText(name).parentElement
    const removeFromCartButton = getByRole(productElement, 'button', {name: "Remove from cart"})
    return removeFromCartButton
  }

  const smileAddToCartButton = getAddToCartButton("smile")
  const frownAddToCartButton = getAddToCartButton("frown")
  const terrorAddToCartButton = getAddToCartButton("terror")

  const smileRemoveFromCartButton = getRemoveFromCartButton("smile")
  const frownRemoveFromCartButton = getRemoveFromCartButton("frown")
  const terrorRemoveFromCartButton = getRemoveFromCartButton("terror")

  userEvent.click(smileAddToCartButton)
  userEvent.click(frownAddToCartButton)
  userEvent.click(terrorAddToCartButton)

  userEvent.click(smileRemoveFromCartButton)
  userEvent.click(frownRemoveFromCartButton)
  userEvent.click(terrorRemoveFromCartButton)

  expect(addToCartFunction).toBeCalledTimes(3)
  expect(addToCartFunction).toBeCalledWith("smile")
  expect(addToCartFunction).toBeCalledWith("frown")
  expect(addToCartFunction).toBeCalledWith("terror")

  expect(removeFromCartFunction).toBeCalledTimes(3)
  expect(removeFromCartFunction).toBeCalledWith("smile")
  expect(removeFromCartFunction).toBeCalledWith("frown")
  expect(removeFromCartFunction).toBeCalledWith("terror")
})

