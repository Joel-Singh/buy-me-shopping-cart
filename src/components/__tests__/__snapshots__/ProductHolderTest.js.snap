// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product holder properly renders a single product 1`] = `
<div>
  <div
    class="product-holder"
  >
    <div
      class="store-product"
    >
      <img
        src="placeholder"
      />
      <button
        aria-label="Remove from cart"
        class="store-product__remove-button"
        type="button"
      >
        --
      </button>
      <span
        aria-label="Amount in cart"
        data-testid="Amount in cart"
      >
        5
      </span>
      <button
        aria-label="Add to cart"
        class="store-product__add-button"
        type="button"
      >
        ++
      </button>
      <div
        class="store-product__name"
      >
        Me smiling
      </div>
    </div>
  </div>
</div>
`;

exports[`Product holder properly renders three products 1`] = `
<div>
  <div
    class="product-holder"
  >
    <div
      class="store-product"
    >
      <img
        src="frown placeholder"
      />
      <button
        aria-label="Remove from cart"
        class="store-product__remove-button"
        type="button"
      >
        --
      </button>
      <span
        aria-label="Amount in cart"
        data-testid="Amount in cart"
      >
        9
      </span>
      <button
        aria-label="Add to cart"
        class="store-product__add-button"
        type="button"
      >
        ++
      </button>
      <div
        class="store-product__name"
      >
        Frown
      </div>
    </div>
    <div
      class="store-product"
    >
      <img
        src="smile placeholder"
      />
      <button
        aria-label="Remove from cart"
        class="store-product__remove-button"
        type="button"
      >
        --
      </button>
      <span
        aria-label="Amount in cart"
        data-testid="Amount in cart"
      >
        2
      </span>
      <button
        aria-label="Add to cart"
        class="store-product__add-button"
        type="button"
      >
        ++
      </button>
      <div
        class="store-product__name"
      >
        Smile
      </div>
    </div>
    <div
      class="store-product"
    >
      <img
        src="depression placeholder"
      />
      <button
        aria-label="Remove from cart"
        class="store-product__remove-button"
        type="button"
      >
        --
      </button>
      <span
        aria-label="Amount in cart"
        data-testid="Amount in cart"
      >
        6
      </span>
      <button
        aria-label="Add to cart"
        class="store-product__add-button"
        type="button"
      >
        ++
      </button>
      <div
        class="store-product__name"
      >
        Depression
      </div>
    </div>
  </div>
</div>
`;
